Blog: Add blog post about DataFusion 50.0.0 release (#115)
apache/datafusion-site: content/blog/2025-09-29-datafusion-50.0.0.md

The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

[TOC]

<!-- see https://github.com/apache/datafusion/issues/16347 for details -->

## Introduction

We are proud to announce the release of [DataFusion 50.0.0]. This blog post
highlights some of the major improvements since the release of [DataFusion
49.0.0]. The complete list of changes is available in the [changelog].

[DataFusion 50.0.0]: https://crates.io/crates/datafusion/50.0.0
[DataFusion 49.0.0]: https://datafusion.apache.org/blog/2025/07/28/datafusion-49.0.0/
[changelog]: https://github.com/apache/datafusion/blob/branch-50/dev/changelog/50.0.0.md

## Performance Improvements 🚀

> **📝TODO** *Update chart*

> 

Reviewer: I updated https://alamb.github.io/datafusion-benchmarking/ and pushed a commit with a new picture: 

<img width="1327" height="530" alt="Screenshot 2025-09-25 at 3 12 23 PM" src="https://github.com/user-attachments/assets/22317ca8-a769-48c5-9524-f47f720a18d4" />

1 row(s) fetched.
Elapsed 0.003 seconds.
```

### `QUALIFY` Clause

The `QUALIFY` clause is now available in DataFusion
([#16933](https://github.com/apache/datafusion/pull/16933)). It allows window
function columns to be filtered without requiring a subquery (similarly to what
`HAVING` does for aggregations).

For example, this query:
```sql
SELECT a, b, c
FROM (
   SELECT a, b, c, rank() OVER(PARTITION BY a ORDER BY b) as rk
   FROM t
)
WHERE rk = 1
```

can now be written like this:
```sql
SELECT a, b, c, rank() OVER(PARTITION BY a ORDER BY b) as rk
FROM t
QUALIFY rk = 1
```

Although it is not a part of the SQL standard (yet), it has been gaining
adoption in several SQL analytical systems, such as DuckDB, Snowflake, and
BigQuery.

### `FILTER` Support for Window Functions

Keeping with the theme, the `FILTER` clause has been extended to support
[aggregate window functions](https://github.com/apache/datafusion/pull/17378).
This allows these functions to be applied to specific rows without having to
rely on `CASE` expressions, similar to what was already possible with regular
aggregate functions.

> > **📝TODO** *Add a practical example?*

Reviewer: I poked around and asked chatGPT and to be honest I couldn't really come up with anything compelling that actually meant something

How about we start with the example from the ticket? 

```suggestion

For example, you could gather multiple distinct sets of  values matching different criteria with a single pass over the input:

```sql
SELECT 
  ARRAY_AGG(c2) FILTER (WHERE c2 >= 2) OVER (...)        -- e.g. [2, 3, 4]
  ARRAY_AGG(CASE WHEN c2 >= 2 THEN c2 END) OVER (...)    -- e.g. [NULL, NULL, 2, 3, 4]
...
FROM table
```

```

1 row(s) fetched.
Elapsed 0.003 seconds.
```

### `QUALIFY` Clause

The `QUALIFY` clause is now available in DataFusion
([#16933](https://github.com/apache/datafusion/pull/16933)). It allows window
function columns to be filtered without requiring a subquery (similarly to what
`HAVING` does for aggregations).

For example, this query:
```sql
SELECT a, b, c
FROM (
   SELECT a, b, c, rank() OVER(PARTITION BY a ORDER BY b) as rk
   FROM t
)
WHERE rk = 1
```

can now be written like this:
```sql
SELECT a, b, c, rank() OVER(PARTITION BY a ORDER BY b) as rk
FROM t
QUALIFY rk = 1
```

Although it is not a part of the SQL standard (yet), it has been gaining
adoption in several SQL analytical systems, such as DuckDB, Snowflake, and
BigQuery.

### `FILTER` Support for Window Functions

Keeping with the theme, the `FILTER` clause has been extended to support
[aggregate window functions](https://github.com/apache/datafusion/pull/17378).
This allows these functions to be applied to specific rows without having to
rely on `CASE` expressions, similar to what was already possible with regular
aggregate functions.

> **📝TODO** *Add a practical example?*

> ### Behavior of User-Defined Functions

Reviewer: ```suggestion
### `ConfigOptions` Now Available to Functions
```

> DataFusion 50.0.0 now allows User-Defined Functions (UDFs) to access the global
> configuration parameters
> ([#16970](https://github.com/apache/datafusion/pull/16970)), allowing their
> behavior to better suit users' workloads. As an example, time UDFs can now use
> custom time zones instead of being limited to UTC.

Reviewer: ```suggestion
DataFusion 50.0.0 passes User-Defined Functions (UDFs) the session 
configuration parameters in [ScalarFunctionArgs](https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.ScalarFunctionArgs.html), see ([#16970](https://github.com/apache/datafusion/pull/16970)). This allows 
behavior that vary based on runtime state such as time UDFs can use
session specified time zones instead in addition to UTC.
```

[aggregate window functions](https://github.com/apache/datafusion/pull/17378).
This allows these functions to apply to specific rows without having to
rely on `CASE` expressions, similar to what was already possible with regular
aggregate functions.

For example, we can gather multiple distinct sets of values matching different
criteria with a single pass over the input:

```sql
SELECT 
  ARRAY_AGG(c2) FILTER (WHERE c2 >= 2) OVER (...)     -- e.g. [2, 3, 4]
  ARRAY_AGG(CASE WHEN c2 >= 2 THEN c2 END) OVER (...) -- e.g. [NULL, NULL, 2, 3, 4]
...
FROM table
```

### `ConfigOptions` Now Available to Functions

DataFusion 50.0.0 now passes session configuration parameters to User-Defined
Functions (UDFs) via
[ScalarFunctionArgs](https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.ScalarFunctionArgs.html)
([#16970](https://github.com/apache/datafusion/pull/16970)). This allows
behavior that varies based on runtime state; for example, time UDFs can use the
session-specified time zone instead of just UTC.

### Additional Apache Spark Compatible Functions

Finally, due to Apache Spark's impact on analytical processing, many DataFusion

> users desire Spark-compatible functionality in their workloads, so DataFusion provides a

Reviewer 2: wordy: "Spark-compatible functionality "

"Spark compatibility"

[mailing list]: https://lists.apache.org/list.html?dev@datafusion.apache.org
[Using External Indexes, Metadata Stores, Catalogs and Caches to Accelerate Queries on Apache Parquet]: https://datafusion.apache.org/blog/2025/08/15/external-parquet-indexes/
[Dynamic Filters: Passing Information Between Operators During Execution for 25x Faster Queries]: https://datafusion.apache.org/blog/2025/09/10/dynamic-filters/
[Implementing User Defined Types and Custom Metadata in DataFusion]: https://datafusion.apache.org/blog/2025/09/21/custom-types-using-metadata/

## New Features ✨

### Improved Spilling Sorts for Larger-than-Memory Datasets

DataFusion has long been able to sort datasets that do not fit entirely in memory,
but still struggled with particularly large inputs or highly memory-constrained 
setups. Larger-than-memory sorts in DataFusion 50.0.0 have been improved with the recent introduction
of multi-level merge sorts (more details in the respective
[ticket](https://github.com/apache/datafusion/pull/15700)). It is now
possible to execute almost any sorting query that would have previously triggered *out-of-memory*

> errors, by relying on disk spilling. Thanks to [Raz Luvaton], [Yongting You], and

Reviewer: fyi @rluvaton, @2010YOUY01 and @ding-young as you are mentioned here

| .../t.parquet | 2025-09-21T17:40:13.650 | 420827020       | 0-63f5331fb4458-19154f8c | NULL    | 44480534            | 27   | page_index=true |
+---------------+-------------------------+-----------------+--------------------------+---------+---------------------+------+-----------------+
1 row(s) fetched.
Elapsed 0.003 seconds.
```

### `QUALIFY` Clause

DataFusion now supports the `QUALIFY` SQL clause
([#16933](https://github.com/apache/datafusion/pull/16933)), which simplifies
filtering window function output (similar to how `HAVING` filters
aggregation output).

For example, filtering the output of the `rank()` function previously
required a query like this:

```sql
SELECT a, b, c
FROM (
   SELECT a, b, c, rank() OVER(PARTITION BY a ORDER BY b) as rk
   FROM t
)
WHERE rk = 1
```

The same query can now be written like this:
```sql
SELECT a, b, c, rank() OVER(PARTITION BY a ORDER BY b) as rk
FROM t
QUALIFY rk = 1
```

Although it is not part of the SQL standard (yet), it has been gaining
adoption in several SQL analytical systems such as DuckDB, Snowflake, and

> BigQuery. Thanks to [Huaijin] and [Jonah Gao] for delivering this feature.

Reviewer: FYI @haohuaijin and @jonahgao as you are mentioned here
